tools/sv-report

```diff
@@ -272,177 +217,430 @@ meta_tags, meta_urls = readConfig(args.meta_tags)
 
 urls = {**urls, **meta_urls}
 
-logger.info("Generating {} from log files in '{}'".format(args.out, args.logs))
 
-data = {}
+@enum.unique
+class TestStatus(enum.Enum):
+    NA = "test-na"
+    PASSED = "test-passed"
+    FAILED = "test-failed"
+    VARIED = "test-varied"
+
+    def __str__(self):
+        return self.value
+
+
+@dataclasses.dataclass
+class TestResult:
+    log_html_file: str = ""
+
+    name: str = ""
+    tags: Set[str] = dataclasses.field(default_factory=set)
+    types: Set[str] = dataclasses.field(default_factory=set)
+    input_files: List[str] = dataclasses.field(default_factory=list)
+    # Unit: bytes
+    total_input_files_size: int = 0
+    status: TestStatus = TestStatus.NA
+    exit_code: int = 0
+
+    # Unit: seconds
+    timeout: int = 0
+
+    # Unit: seconds
+    total_time: float = 0
+    user_time: float = 0
+    system_time: float = 0
+
+    # Unit: KB
+    ram_usage: float = 0
+
+
+@dataclasses.dataclass
+class TagResults:
+    status: TestStatus = TestStatus.NA
+    types: List[str] = dataclasses.field(default_factory=list)
+    passed_tests: int = 0
+    tests: List[TestResult] = dataclasses.field(default_factory=list)
+
+
+@dataclasses.dataclass
+class ToolResults:
+    tests: Dict[str, TestResult] = dataclasses.field(default_factory=dict)
+    tags: Dict[str, TagResults] = dataclasses.field(default_factory=dict)
+
+    # Unit: seconds
+    total_time: float = 0
+    user_time: float = 0
+    system_time: float = 0
+
+    # Unit: MB
+    max_ram_usage: float = 0
+
+    # Unit: KiB/s
+    passed_throughput: float = 0
+
+    # Tool name. This should be used only for display purposes. Always use
+    # a key from `tools` in `ReportResults` as a runner/tool ID.
+    name: str = ""
+    version: str = ""
+    url: str = ""
+
+    total_passed_tests: int = 0
+    total_passed_tags: int = 0
+    total_tested_tags: int = 0
+
+    @property
+    def total_tests(self):
+        return len(self.tests)
+
+
+@dataclasses.dataclass
+class TagInfo:
+    description: str = ""
+    url: str = ""
+
+
+@dataclasses.dataclass
+class ReportResults:
+    # Results from each runner/tool.
+    # Keys are runner names (i.e. runner class names)
+    tools: Dict[str, ToolResults] = dataclasses.field(default_factory=dict)
+    # Tags used in the report. Keys are tag IDs.
+    tags: Dict[str, TagInfo] = dataclasses.field(default_factory=dict)
+
+
+# Data passed to report template
+report_data = ReportResults()
+
+PATH_WITH_LINE_NO_MATCHER = re.compile(
+    r"([^\s:\"'`/]*/[^\s:\"'`]+\.\w+)(:(\d+)(?::\d+)?)?")
+
 
-for tag in database:
-    tag_usage[tag] = False
+def convertPathsToRelativeLinks(text: str, relative_to: str):
```

Comment:
It's not clear what it does, please explain, including the regex pattern

Reply:
Documented.